core/rawdb/chain_iterator.go

```diff
@@ -28,157 +28,164 @@ import (
+			logged = time.Now()
+		}
 	}
-	abort := make(chan struct{})
-	defer close(abort)
+	if err := batch.Write(); err != nil {
+		log.Crit("Failed to write data to db", "err", err)
+	}
+	batch.Reset()
 
-	threads := to - from
-	if cpus := runtime.NumCPU(); threads > uint64(cpus) {
-		threads = uint64(cpus)
+	WriteHeadHeaderHash(db, hash)
+	WriteHeadFastBlockHash(db, hash)
+	log.Info("Initialized database from freezer", "blocks", frozen, "tail", frozen, "elapsed", common.PrettyDuration(time.Since(start)))
+}
+
+type blockTxHashes struct {
+	number uint64
+	hashes []common.Hash
+}
+
+// iterateTransactions iterates over all transactions in the (canon) block
+// number(s) given, and yields the hashes on a channel
+func iterateTransactions(db ethdb.Database, from uint64, to uint64, reverse bool) (chan *blockTxHashes, chan struct{}) {
+	// One thread sequentially reads data from db
+	type numberRlp struct {
+		number uint64
+		rlp    rlp.RawValue
 	}
-	for i := 0; i < int(threads); i++ {
-		go func() {
-			for {
-				// Fetch the next task number, terminating if everything's done
-				var n int64
-				if !reverse {
-					n = atomic.AddInt64(&number, 1)
-					if n >= int64(to) {
-						return
-					}
-				} else {
-					n = atomic.AddInt64(&number, -1)
-					if n < int64(from) {
-						return
-					}
-				}
-				block := ReadBlock(db, ReadCanonicalHash(db, uint64(n)), uint64(n))
-				if prepare != nil && block != nil {
-					prepare(block)
-				}
-				// Feed the block to the aggregator, or abort on interrupt
-				select {
-				case results <- block:
-				case <-abort:
-					return
-				}
-			}
-		}()
+	if to == from {
+		return nil, nil
 	}
-	// Reassemble the blocks into a contiguous stream and apply the action callback
 	var (
-		next, first, last int64
-		queue             = prque.New(nil)
-
-		batch  = db.NewBatch()
-		start  = time.Now()
-		logged = start.Add(-7 * time.Second) // Unindex during import is fast, don't double log
+		threads  = to - from
+		rlpCh    = make(chan *numberRlp, threads*2)     // we send raw rlp over this channel
+		hashesCh = make(chan *blockTxHashes, threads*2) // send hashes over hashesCh
+		abortCh  = make(chan struct{})
 	)
-	if !reverse {
-		next, first, last = int64(from), int64(from), int64(to)
-	} else {
-		next, first, last = int64(to-1), int64(to-1), int64(from-1)
+	if cpus := runtime.NumCPU(); threads > uint64(cpus) {
+		threads = uint64(cpus)
 	}
-	for i := from; i < to; i++ {
-		// Retrieve the next result and bail if it's nil
-		block := <-results
-		if block == nil {
-			return errors.New("broken database")
-		}
-		// Push the block into the import queue and process contiguous ranges
-		priority := -int64(block.NumberU64())
+	// lookup runs in one instance
+	lookup := func() {
+		n, end := from, to
 		if reverse {
-			priority = int64(block.NumberU64())
+			n, end = to-1, from-1
```

Comment:
`from` can be 0 here. So potential underflow

Reply:
Heh, I added a testcase to catch that, and the thing is, yes, it oveflows.. but so does `n--` further down, so the equality check `for n != end` still works as it ought to